### PR TITLE
 ✨(project) add a production-ready compose file

### PR DESCRIPTION
**Pull request by wilbrdt**
This adds a production-read compose file, to be able to deploy warren-tdbp with Docker Swarm.

### CODE REVIEW

File: production/docker-compose.prod.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+services:
+  postgresql:
+    image: postgres:12
+    env_file:
+      - env/postgresql.env
+    volumes:
+      - postgres_data:/var/lib/postgresql/data/
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready -U postgres"]
+      interval: 1s
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]: Maybe a bit too aggressive?

File: production/docker-compose.prod.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      retries: 60
+    deploy:
+      replicas: 1
+      restart_policy:
+        condition: on-failure
+  
+  app:
+    # image: apuiavignonuniversity/warren-tdbp:app-0.1.0
+    image: warren-tdbp:app-0.1.0
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]:
```suggestion
    image: apuiavignonuniversity/warren-tdbp:app-0.1.0
```

File: production/docker-compose.prod.yml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    environment:
+      - WARREN_APP_MEDIA_ROOT=${WARREN_APP_MEDIA_ROOT}
+      - WARREN_APP_STATIC_ROOT=${WARREN_APP_STATIC_ROOT}
+    env_file:
+      - env/app.env
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml:24]: Maybe we should provide example dotenv files ready-for production?

File: production/docker-compose.prod.yml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      replicas: 1
+      restart_policy:
+        condition: on-failure
+  
+  app:
+    # image: apuiavignonuniversity/warren-tdbp:app-0.1.0
+    image: warren-tdbp:app-0.1.0
+    environment:
+      - WARREN_APP_MEDIA_ROOT=${WARREN_APP_MEDIA_ROOT}
+      - WARREN_APP_STATIC_ROOT=${WARREN_APP_STATIC_ROOT}
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]: IMO we should not rely on defaults set from the environment. And those should be set in the dotenv file.

File: production/docker-compose.prod.yml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    ports:
+      - "8000:8000"
+    command:
+      - "gunicorn"
+      - "-c"
+      - "/usr/local/etc/gunicorn/warren.py"
+      - "warren.wsgi:application"
+    volumes:
+      - media:${WARREN_APP_MEDIA_ROOT}
+      - static:${WARREN_APP_STATIC_ROOT}
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]: Same remark for defaults.

File: production/docker-compose.prod.yml [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      retries: 3
+      start_period: 15s
+    deploy:
+      replicas: 3
+      restart_policy:
+        condition: on-failure
+
+  api:
+    # image: apuiavignonuniversity/warren-tdbp:api-0.1.0
+    image: warren-tdbp:api-0.1.0
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]:
```suggestion
    image: apuiavignonuniversity/warren-tdbp:api-0.1.0
```

File: production/docker-compose.prod.yml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    env_file:
+      - env/api.env
+    ports:
+      - "${WARREN_API_SERVER_PORT:-8100}:${WARREN_API_SERVER_PORT:-8100}"
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]:
```suggestion
    ports:
      - "8100:8100"
```

File: production/docker-compose.prod.yml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    command:
+      - uvicorn
+      - "warren.api:app"
+      - "--proxy-headers"
+      - "--log-config"
+      - "core/logging_config.yaml"
+      - "--host"
+      - "0.0.0.0"
+      - "--port"
+      - "${WARREN_API_SERVER_PORT:-8100}"
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]:
```suggestion
      - "8100"
```

File: production/docker-compose.prod.yml [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      - "--reload"
+    configs:
+      - source: logging_config.yaml
+        target: /app/core/logging_config.yaml
+    healthcheck:
+      test: curl -f http://localhost:${WARREN_API_SERVER_PORT}/__heartbeat__ || exit 1
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml]: Not sure when this variable will be interpreted :thinking:

File: production/docker-compose.prod.yml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      - "--log-config"
+      - "core/logging_config.yaml"
+      - "--host"
+      - "0.0.0.0"
+      - "--port"
+      - "${WARREN_API_SERVER_PORT:-8100}"
+      - "--reload"
+    configs:
+      - source: logging_config.yaml
+        target: /app/core/logging_config.yaml
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml:65]: We should document this part.

File: production/docker-compose.prod.yml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  logging_config.yaml:
+    external: true
+
+volumes:
+  postgres_data:
+    external: true
+  media:
+    external: true
+  static:
+    external: true
```
Comment (jmaupetit) [inline production/docker-compose.prod.yml:87]: We should also document how to create volumes.

**Review by jmaupetit — COMMENTED** [review]
Looks great! Maybe we can write a temporary documentation while it's hot before waiting to integrate it in the project's docs. WDYT?

File: production/docker-compose.prod.yml (hunk 4 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml]: Fixed!

File: production/docker-compose.prod.yml (hunk 9 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml]: Removed!

File: production/docker-compose.prod.yml (hunk 10 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml:65]: Documented briefly in the `README.md`

File: production/docker-compose.prod.yml (hunk 11 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml:87]: Documented briefly on the `README.md`

File: production/docker-compose.prod.yml (hunk 3 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml:24]: Whoops, I forget to version them :grin:

File: production/docker-compose.prod.yml (hunk 1 above)
Comment (wilbrdt) [reply production/docker-compose.prod.yml]: Indeed! Fixed

**Comment by wilbrdt** [comment]
Added a very brief documentation, will need to be expanded

File: production/README.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+This quick tutorial aims at deploying Warren-TdBP on a local Docker Engine cluster in
+swarm mode. In this tutorial, you will learn to:
+
+- run and configure a small cluster of Docker Engines in swarm mode on your machine,
+- deploy Warren TdBP from the production Compose file.
+
+Initialize a local cluster with the command:
+```bash
+ docker swarm init
+```
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
```bash
docker swarm init
```
```

File: production/README.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+```bash
+docker info
+```
+
+We can view information about nodes with:
+```bash
+docker node ls
+```
+
+> In this tutorial, we'll deploy Warren-TdBP on a single node cluster. If you want to
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
> In this tutorial, we will deploy Warren-TdBP on a single node cluster. If you want to
```

File: production/README.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+> add more nodes to the cluster, refer to this
+> [documentation](https://docs.docker.com/engine/swarm/swarm-tutorial/add-nodes/).
+
+### Creating the volumes
+
+Warren-TdBP is a Django application running with a PostgreSQL database. We need to
+create a Docker volume for the database and for the static and media files.
+
+In this tutorial, we'll create three simple volumes:
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
In this tutorial, we'll create three volumes:
```

File: production/README.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+- run and configure a small cluster of Docker Engines in swarm mode on your machine,
+- deploy Warren TdBP from the production Compose file.
+
+Initialize a local cluster with the command:
+```bash
+ docker swarm init
+```
+
+We can view the current state of the swarm with:
```
Comment (jmaupetit) [inline production/README.md:20]: Should we say "the swarm" or "the swarm orchestrator"?

File: production/README.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+docker node ls
+```
+
+> In this tutorial, we'll deploy Warren-TdBP on a single node cluster. If you want to
+> add more nodes to the cluster, refer to this
+> [documentation](https://docs.docker.com/engine/swarm/swarm-tutorial/add-nodes/).
+
+### Creating the volumes
+
+Warren-TdBP is a Django application running with a PostgreSQL database. We need to
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
Warren-TdBP `app` and `api` services need a PostgreSQL database for persistency. We need to
```

File: production/README.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+create a Docker volume for the database and for the static and media files.
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
create a Docker volume for the database, but also for the static and media files (required by the `app` service).
```

File: production/README.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+In this tutorial, we'll create three simple volumes:
+```bash
+docker volume create postgresql_data
+docker volume create media
+docker volume create static
+```
+
+> For more options on the volume creation, you can refer to this
+> [documentation](https://docs.docker.com/engine/reference/commandline/volume_create/).
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
> For more options on the volume creation (size, file system, device, etc.), you can refer to this
> [documentation](https://docs.docker.com/engine/reference/commandline/volume_create/).
```

File: production/README.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+### Creating the logging config
+
+We may want to adjust logging levels to debug our application after the deployment.
+Let's create a config from the logging configuration file:
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion 
To adjust services logging configuration, you can create  a new `config` object from the project's sources:
```

File: production/README.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+
+```bash
+docker config create logging_config.yaml src/api/logging-config.prod.yaml
+```
```
Comment (jmaupetit) [inline production/README.md:58]:
```suggestion
```

In this example, we will create a new `config` called `logging_config.yaml` from the content of the `src/api/logging-config.prod.yaml` file. This file will be mounted in `x/y/z` during the deployment.
```

File: production/README.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+And we can see all the running containers with:
+```bash
+docker ps
+```
+
+If a service is not healthy, we can investigate further with:
+```bash
+docker inspect --format "{{json .State.Health }}" CONTAINER_ID | jq
+```
```
Comment (jmaupetit) [inline production/README.md:91]:
```suggestion
```
_Nota bene_: this command requires that [jq](https://jqlang.github.io/jq/) is installed in your operating system.
```

File: production/README.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+docker config create logging_config.yaml src/api/logging-config.prod.yaml
+```
+
+### Deploying Warren-TdBP
+
+Docker Engine in swarm mode can deploy services defined in a Compose file. We are going
+to deploy three services: `postgresql`, `api` and `app`. 
+
+You can adjust the `docker-compose.prod.yml` file to fit your needs, along with the
+three environment files: `postresql.env`, `api.env` and `app.env`.
```
Comment (jmaupetit) [inline production/README.md]:
```suggestion
You can adjust the `docker-compose.prod.yml` file to fit your needs, along with the
three environment files: `postresql.env`, `api.env` and `app.env`. Pay attention to 
Warren's docker images tags.
```

**Review by jmaupetit — APPROVED** [review]
I only have minor remarks on the documentation :sunglasses:

File: production/README.md (hunk 15 above)
Comment (wilbrdt) [reply production/README.md:20]: Docker uses `the swarm` in their [documentation](https://docs.docker.com/engine/swarm/swarm-tutorial/create-swarm/), going with that!

**Pull request merged**